File: src/main/java/me/testaccount666/serversystem/commands/executables/enderchest/offline/ListenerOfflineEnderChest.java
package me.testaccount666.serversystem.commands.executables.enderchest.offline;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.annotations.RequiredCommands;
import me.testaccount666.serversystem.commands.executables.enderchest.CommandEnderChest;
import me.testaccount666.serversystem.commands.interfaces.ServerSystemCommandExecutor;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.event.player.PlayerLoginEvent;

import java.util.ArrayList;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

@RequiredCommands(requiredCommands = CommandEnderChest.class)
public class ListenerOfflineEnderChest implements Listener {
    private EnderChestLoader _enderChestLoader;
    private CommandEnderChest _enderChest;

    public boolean canRegister(Set<ServerSystemCommandExecutor> requiredCommands) {
        var canRegister = new AtomicBoolean(false);

        requiredCommands.forEach(command -> {
            if (!(command instanceof CommandEnderChest commandEnderChest)) return;
            var enderChestLoader = commandEnderChest.offlineEnderChest.enderChestLoader;
            if (enderChestLoader == null) return;

            _enderChest = commandEnderChest;
            _enderChestLoader = enderChestLoader;
            canRegister.set(true);
        });

        return canRegister.get();
    }

    @EventHandler
    public void onViewedJoin(PlayerLoginEvent event) {
        var viewedPlayer = event.getPlayer();

        var inventoryOptional = _enderChestLoader.inventoryMap.getValue(viewedPlayer.getUniqueId());
        if (inventoryOptional.isEmpty()) {
            Bukkit.getLogger().info("OfflineEnderChest: No offline inventory found for player ${viewedPlayer.getName()}");
            return;
        }
        var inventory = inventoryOptional.get();

        Bukkit.getLogger().info("OfflineEnderChest: Offline inventory was found for player ${viewedPlayer.getName()}");

        var viewers = new ArrayList<>(inventory.getViewers());
        inventory.close();

        _enderChestLoader.inventoryMap.removeByKey(viewedPlayer.getUniqueId());
        _enderChestLoader.saveOfflineInventory(viewedPlayer.getUniqueId(), inventory);

        Bukkit.getScheduler().runTaskLater(ServerSystem.Instance, () -> viewers.forEach(viewer -> {
            if (!(viewer instanceof Player player)) return;

            var optionalUser = ServerSystem.Instance.getUserManager().getUser(player);
            if (optionalUser.isEmpty()) return;

            var cachedUser = optionalUser.get();
            if (cachedUser.isOfflineUser()) return;
            var user = (User) cachedUser.getOfflineUser();

            _enderChest.executeEnderChestCommand(user, viewedPlayer.getName());
        }), 10L);
    }

    @EventHandler
    public void onInventoryClose(InventoryCloseEvent event) {
        var inventory = event.getInventory();

        if (!_enderChestLoader.inventoryMap.containsValue(inventory)) return;

        var uuid = _enderChestLoader.inventoryMap.getKey(inventory);
        if (uuid.isEmpty()) return;

        _enderChestLoader.inventoryMap.removeByValue(inventory);
        _enderChestLoader.saveOfflineInventory(uuid.get(), inventory);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/enderchest/offline/ListenerOfflineEnderChest.kt
package me.testaccount666.serversystem.commands.executables.enderchest.offline

import me.testaccount666.paperktx.scheduler.skedule.okkero.schedule
import me.testaccount666.serversystem.ServerSystem.Companion.instance
import me.testaccount666.serversystem.ServerSystem.Companion.log
import me.testaccount666.serversystem.annotations.RequiredCommands
import me.testaccount666.serversystem.commands.executables.enderchest.online.CommandEnderChest
import me.testaccount666.serversystem.commands.interfaces.ServerSystemCommandExecutor
import me.testaccount666.serversystem.extensions.getService
import me.testaccount666.serversystem.userdata.UserManager
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.player.PlayerLoginEvent

@RequiredCommands([CommandEnderChest::class])
class ListenerOfflineEnderChest : Listener {
    private lateinit var _enderChestLoader: EnderChestLoader
    private lateinit var _enderChest: CommandEnderChest

    fun canRegister(requiredCommands: Set<ServerSystemCommandExecutor>): Boolean {
        var canRegister = false

        requiredCommands.forEach {
            _enderChest = it as? CommandEnderChest ?: return@forEach
            _enderChestLoader = it.offlineEnderChest.enderChestLoader ?: return@forEach

            canRegister = true
        }

        return canRegister
    }

    @EventHandler
    fun onViewedJoin(event: PlayerLoginEvent) {
        val viewedPlayer = event.getPlayer()

        val inventory = _enderChestLoader.inventoryMap.getValue(viewedPlayer.uniqueId) ?: run {
            log.fine("OfflineEnderChest: No offline inventory found for player ${viewedPlayer.name}")
            return
        }

        log.fine("OfflineEnderChest: Offline inventory was found for player ${viewedPlayer.name}")

        val viewers = ArrayList(inventory.viewers)
        inventory.close()

        _enderChestLoader.inventoryMap.removeByKey(viewedPlayer.uniqueId)
        _enderChestLoader.saveOfflineInventory(viewedPlayer.uniqueId, inventory)

        instance.schedule {
            waitFor(10L)
            viewers.filterIsInstance<Player>().forEach { viewer ->
                val cachedUser = getService<UserManager>().getUserOrNull(viewer) ?: return@forEach

                if (cachedUser.isOfflineUser) return@forEach
                val user = cachedUser.onlineUser
                _enderChest.executeEnderChestCommand(user, viewedPlayer.name)
            }
        }
    }

    @EventHandler
    fun onInventoryClose(event: InventoryCloseEvent) {
        val inventory = event.inventory
        val uuid = _enderChestLoader.inventoryMap.getKey(inventory) ?: return

        _enderChestLoader.inventoryMap.removeByValue(inventory)
        _enderChestLoader.saveOfflineInventory(uuid, inventory)
    }
}
